### ISSUE TITLE
Not Mirroring roles.

### ISSUE THREAD

**Issue by MalikZohaibMustafa** [crimilo/Discord-Mirror#21]
"replacements": {
   "content": [
      {
         "replace": "ORIGINAL_ROLE_OR_USER_OR_CHANNEL_ID",
         "with": "REPLACED_ROLE_OR_USER_OR_CHANNEL_ID"
      }
   ]
}

I tries this part. 
aSSUME THERE IS A Source_Server and I am a member(not admin) in that source server. I used my auth token , and channel id of that source server inn that script, Now in my secord account there is the destination server and I am admin of it. (Note that 2nd account is not part of source server). So i have used webhook of dest server in the script too.

The problem is that in source server there are many roles in the server and whichever is called, it doestnot show them in dest server.
instead it shows @deleted-role.

 I have tried below scripts .
  1. 
  "replacements": {
            "content": [
               {
                  "replace": "Admin",
                  "with": "Dyno"
               }
            ]

  2. 
  "replacements": {
            "content": [
               {
                  "replace": "@Admin",
                  "with": "@Dyno"
               }
            ]
  3. 
  "replacements": {
            "content": [
               {
                  "replace": "1087705203484135475",
                  "with": "@Dyno"
               }
            ]

            but still it shows deleted role no matter what I try. Please help me resolve it. I only have role ids of dest servers. No matter what role comes in source channel, I want to display my specified role in each message. Note i have multiple channels and roles. So role depends upon channel  wise. I want to provide role id in each mirror to change the text @deleted-role to that role.

**Comment by crimilo** [comment]
> "replacements": {
> 
>    "content": [
> 
>       {
> 
>          "replace": "ORIGINAL_ROLE_OR_USER_OR_CHANNEL_ID",
> 
>          "with": "REPLACED_ROLE_OR_USER_OR_CHANNEL_ID"
> 
>       }
> 
>    ]
> 
> }
> 
> 
> 
> I trieD this part. 
> 
> Assume there IS A Source_Server and I am a member(not admin) in that source server. I used my auth token , and channel id of that source server inn that script, Now in my second account there is the destination server and I am admin of it. (Note that 2nd account is not part of source_server). So i have used webhook of destination server in the script too.
> 
> 
> 
> The problem is that in source server there are many roles in the server and whichever is called, it doestnot show them in dest server.
> 
> instead it shows @deleted-role.
> 
> 
> 
>  I have tried below scripts .
> 
>   1. 
> 
>   "replacements": {
> 
>             "content": [
> 
>                {
> 
>                   "replace": "Admin",
> 
>                   "with": "Dyno"
> 
>                }
> 
>             ]
> 
> 
> 
>   2. 
> 
>   "replacements": {
> 
>             "content": [
> 
>                {
> 
>                   "replace": "@Admin",
> 
>                   "with": "@Dyno"
> 
>                }
> 
>             ]
> 
>   3. 
> 
>   "replacements": {
> 
>             "content": [
> 
>                {
> 
>                   "replace": "1087705203484135475",
> 
>                   "with": "@Dyno"
> 
>                }
> 
>             ]
> 
> 
> 
> 
> 
>             but still it shows deleted role no matter what I try. Please help me resolve it. 
> 
>             
> 
>             I only have role ids of roles in dest servers. No matter what role comes in source channel, I want to display my specified role in each message. Note i have multiple channels and roles. So role depends upon channel  wise. I want to provide role id in each mirror to change the text @deleted-role to that role.
> 
>             
> 
>             
> 
>             
> 
>             
> 
>             

You need to replace the role IDs. For instance, if @admin has ID: 123 and @dyno has ID: 456, to replace @admin with @dyno, you need to configure the bot as follows:

'''
"content": [
   {
      "replace": "123",
      "with": "456"
   }
]
'''

**Comment by MalikZohaibMustafa** [comment]
Acknowledged, but I can't access the role ID of roles in source server because I am not admin of source server.

**Comment by crimilo** [comment]
> Acknowledged, but I can't access the role ID of roles in source server because I am not admin of source server.

You don't need to be admin. Just right-click on the role from a user's profile who has it and click **Copy Role ID**.

**Comment by crimilo** [comment]
If you don't have access to the role id, then you can't replace it.

**Comment by MalikZohaibMustafa** [comment]
Yes brother. I know that. That's what I am asking you.

In my destination server if any message has @deleted-role, The message replacement in your script can remove that "@deleted-role" if found in message and then display the role which I want over there using role id in that place?

**Comment by crimilo** [comment]
No, the bot cannot do that as the mention of "@deleted-role" is not part of the message content. The Discord client automatically generates this mention when it cannot find a valid role.

**Comment by MalikZohaibMustafa** [comment]
Roger That. Thanks Buddy.

**Comment by MalikZohaibMustafa** [comment]
Possible to just remove the @deleted-role from message?

**Comment by MalikZohaibMustafa** [comment]
hey,just a heads up. I modify the script now its working for roles. We just have to give role id of roles in destination server. (one roleid per cchannel). Do you want me to push rthat code?

overiew:

Badically in messagereplacement.js, I modified the function named processMessage with below code:


processMessage(message) {
        const mirrors = require('../config.json').mirrors;
        for (const mirror of mirrors) {
                if (mirror.channel_ids.includes(message.channelId)) {
                const newRoleId = mirror.replacements.content[0].new_role_id;
                this.replaceRole(message, 'mentions', newRoleId);
                break;
            }
        }
        for (const embed of message.embeds) {
            this.processEmbed(embed);
        }
    }
    
    
    also created a new function in messagerelacement.js which is given below.
    
    replaceRole(object,attribute,replacement)
    {
        if (attribute === 'mentions') {
            const mentionString = object.content;
            const roleID = mentionString.match(/\d+/)[0];
            if (object.content.includes(roleID)) {
                object.content = object.content.replace(roleID, replacement);    
               }
        }
    }
    
    and my config.js is:
    
    
     "mirrors": [
       {
           "channel_ids": [ "1039xxxxxxxx9941063" ],
            "webhook_urls": ["https://discord.com/api/webhooks/109578xxxxxx4309241/p8AR8xxxxxx73PZElhnoZ4-f0DHHuSxxxxxxYg8j8kSXPi9m"],
         "requirements": {
            "min_embeds_count": 0,
            "min_content_length": 0,
            "min_attachments_count": 0
         },
         "options": {
            "use_webhook_profile": false,
            "remove_attachments": false,
            "mirror_messages_from_bots": true,
            "mirror_reply_messages": true
         },
          "replacements": {
            "content": [
               {  
                  "new_role_id": "108xxxxxx484135475"
               }
            ]
         }
       }   ]
}
